File: packages/agent-server/src/execution/actions.ts
import type { Action, A11yNode, Execution } from "../ia/types.js";
import { querySelector } from "../ia/selectors.js";
import { execCommand } from "../lib/exec.js";

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export interface ActionContext<TParams = unknown, TPlanState = unknown> {
  a11y: A11yNode;
  screenshot: string; // base64
  execution: Execution<TParams, TPlanState>;
  frame?: A11yNode; // Scoped frame for querySelector (from identify metadata)
}

/**
 * Execute an action.
 *
 * Actions are predefined operations like click, type, scroll, etc.
 * The selector in actions uses CSS-like syntax.
 */
export async function executeAction<TParams, TPlanState = unknown>(
  action: Action,
  ctx: ActionContext<TParams, TPlanState>
): Promise<void> {
  const { a11y, execution, frame } = ctx;
  const session = execution.context.session;
  // Use frame for scoped queries when available, fallback to full a11y tree
  const queryRoot = frame ?? a11y;

  switch (action.type) {
    case "click": {
      let x: number;
      let y: number;

      // Check if coordinates provided directly
      if (action.x !== undefined && action.y !== undefined) {
        x = action.x;
        y = action.y;
      } else if (action.selector) {
        // Find element using CSS-like selector (scoped to frame if available)
        const element = querySelector(queryRoot, action.selector);
        if (!element?.bounds) {
          throw new Error(`Element not found: ${action.selector}`);
        }
        // Calculate center of element
        x = Math.round(element.bounds.x + element.bounds.width / 2);
        y = Math.round(element.bounds.y + element.bounds.height / 2);
      } else {
        throw new Error("Click action requires either selector or x,y coordinates");
      }

      // Build click args - use window activation if frame info available
      const args: string[] = [];
      if (frame?.window?.pid && frame?.bounds) {
        // Activate specific window by PID + bounds before clicking
        args.push(
          "--window",
          String(frame.window.pid),
          String(frame.bounds.x),
          String(frame.bounds.y),
          String(frame.bounds.width),
          String(frame.bounds.height),
          "--"
        );
      }
      args.push(String(x), String(y));
      await execCommand("click", args, { session });
      break;
    }

    case "type": {
      // If selector provided, click first
      if (action.selector) {
        await executeAction({ type: "click", selector: action.selector }, ctx);
        await sleep(100);
      }
      // Type text using clipboard (Unicode-safe)
      await execCommand("type", [action.text], { session });
      break;
    }

    case "key": {
      // Focus window first if frame info available, then press key combo
      const keyArgs: string[] = [];
      if (frame?.window?.pid && frame?.bounds) {
        // Activate specific window by PID + bounds before key press
        keyArgs.push(
          "--window",
          String(frame.window.pid),
          String(frame.bounds.x),
          String(frame.bounds.y),
          String(frame.bounds.width),
          String(frame.bounds.height),
          "--"
        );
      }
      keyArgs.push(action.combo);
      await execCommand("key", keyArgs, { session });
      break;
    }

    case "scroll": {
      // Scroll in direction at optional coordinates
      const args = [action.direction, String(action.amount ?? 3)];
      if (action.x !== undefined && action.y !== undefined) {
        args.push(String(action.x), String(action.y));
      }
      await execCommand("scroll", args, { session });
      break;
    }

    case "wait": {
      await sleep(action.ms);
      break;
    }

    case "emit": {
      execution.emit(action.event);
      break;
    }

    case "sequence": {
      for (const subAction of action.actions) {
        await executeAction(subAction, ctx);
      }
      break;
    }
  }
}

File: packages/agent-server-rust/src/execution/actions.rs
use crate::ia::selectors::query_selector;
use crate::ia::types::{A11yNode, Action, SubscriptionEvent};
use crate::tools::exec::{exec_command, ExecOptions};
use std::future::Future;
use std::pin::Pin;

/// Find the first frame node with window info in the a11y tree.
/// Used for window activation before clicking/typing.
fn find_frame_with_window(node: &A11yNode) -> Option<&A11yNode> {
    if node.role == "frame" && node.window.is_some() && node.bounds.is_some() {
        return Some(node);
    }
    if let Some(children) = &node.children {
        for child in children {
            if let Some(found) = find_frame_with_window(child) {
                return Some(found);
            }
        }
    }
    None
}

/// Build --window activation args from a frame node.
fn window_activate_args(frame: &A11yNode) -> Vec<String> {
    if let (Some(win), Some(fb)) = (&frame.window, &frame.bounds) {
        vec![
            "--window".to_string(),
            win.pid.to_string(),
            (fb.x as i32).to_string(),
            (fb.y as i32).to_string(),
            (fb.width as i32).to_string(),
            (fb.height as i32).to_string(),
            "--".to_string(),
        ]
    } else {
        vec![]
    }
}

/// Execute a single action against the WeChat UI.
/// Returns a BoxFuture to support recursive calls (Sequence action).
pub fn execute_action<'a>(
    action: &'a Action,
    options: &'a ExecOptions,
    a11y: &'a A11yNode,
    emit: &'a (dyn Fn(SubscriptionEvent) + Send + Sync),
) -> Pin<Box<dyn Future<Output = ()> + Send + 'a>> {
    Box::pin(async move {
        match action {
            Action::ClickSelector { selector } => {
                // Find frame for window activation and scoped queries
                let frame = find_frame_with_window(a11y);
                let query_root = frame.unwrap_or(a11y);

                if let Some(node) = query_selector(query_root, selector) {
                    if let Some(bounds) = &node.bounds {
                        let cx = (bounds.x + bounds.width / 2.0).round() as i32;
                        let cy = (bounds.y + bounds.height / 2.0).round() as i32;
                        tracing::info!("[action] click selector '{selector}' → ({cx}, {cy})");

                        let mut args = frame
                            .map(|f| window_activate_args(f))
                            .unwrap_or_default();
                        args.push(cx.to_string());
                        args.push(cy.to_string());
                        let args_ref: Vec<&str> = args.iter().map(|s| s.as_str()).collect();
                        exec_command("click", &args_ref, options).await;
                    } else {
                        tracing::warn!("[action] click selector '{selector}' matched but no bounds");
                    }
                } else {
                    tracing::warn!("[action] click selector '{selector}' — no match");
                }
            }

            Action::ClickCoords { x, y } => {
                let frame = find_frame_with_window(a11y);
                let mut args = frame
                    .map(|f| window_activate_args(f))
                    .unwrap_or_default();
                args.push((*x as i32).to_string());
                args.push((*y as i32).to_string());
                let args_ref: Vec<&str> = args.iter().map(|s| s.as_str()).collect();
                exec_command("click", &args_ref, options).await;
            }

            Action::Type { text, selector: _ } => {
                exec_command("input", &[text.as_str()], options).await;
            }

            Action::Key { combo } => {
                let frame = find_frame_with_window(a11y);
                let mut args = frame
                    .map(|f| window_activate_args(f))
                    .unwrap_or_default();
                args.push(combo.clone());
                let args_ref: Vec<&str> = args.iter().map(|s| s.as_str()).collect();
                exec_command("key", &args_ref, options).await;
            }

            Action::Scroll {
                direction,
                x: _,
                y: _,
                amount,
            } => {
                let dir = match direction {
                    crate::ia::types::ScrollDirection::Up => "up",
                    crate::ia::types::ScrollDirection::Down => "down",
                };
                let mut args = vec![dir.to_string()];
                if let Some(amt) = amount {
                    args.push(amt.to_string());
                }
                let args_ref: Vec<&str> = args.iter().map(|s| s.as_str()).collect();
                exec_command("scroll", &args_ref, options).await;
            }

            Action::Wait { ms } => {
                tokio::time::sleep(std::time::Duration::from_millis(*ms)).await;
            }

            Action::Emit { event } => {
                emit(event.clone());
            }

            Action::Sequence { actions } => {
                for a in actions {
                    execute_action(a, options, a11y, emit).await;
                }
            }
        }
    })
}
